### PR TITLE
Update DictRows and DictColumns definition

### LINKED ISSUE #309: Definition of DictColumns
@quinnj in:
```
const DictColumns = AbstractDict{K, V} where {K <: Union{Symbol, String}, V <: AbstractVector}
```
we require string to be `String` (and not `AbstractString`). What is the rationale behind this?

CC @nalimilan @ablaom

### PR DESCRIPTION
**Pull request by bkamins**
Fixes https://github.com/JuliaData/Tables.jl/issues/309

### CODE REVIEW

File: src/Tables.jl [hunk 1]
```diff
@@ -144,13 +144,13 @@ getcolumn(x::AbstractDict{Symbol}, nm::Symbol) = x[nm]
...
 columnnames(x::AbstractDict{Symbol}) = collect(keys(x))
 
-getcolumn(x::AbstractDict{String}, i::Int) = x[String(columnnames(x)[i])]
-getcolumn(x::AbstractDict{String}, nm::Symbol) = x[String(nm)]
-getcolumn(x::AbstractDict{String}, ::Type{T}, i::Int, nm::Symbol) where {T} = x[String(nm)]
-columnnames(x::AbstractDict{String}) = collect(Symbol(k) for k in keys(x))
+getcolumn(x::AbstractDict{<:AbstractString}, i::Int) = x[String(columnnames(x)[i])]
+getcolumn(x::AbstractDict{<:AbstractString}, nm::Symbol) = x[String(nm)]
+getcolumn(x::AbstractDict{<:AbstractString}, ::Type{T}, i::Int, nm::Symbol) where {T} = x[String(nm)]
+columnnames(x::AbstractDict{<:AbstractString}) = collect(Symbol(k) for k in keys(x))
```
Comment (bkamins) [inline src/Tables.jl:150]: Note that this changes the following (this is the previous behavior):
```
julia> Tables.columnnames(Dict("a"=>1, SubString("b")=>2))
(:slots, :keys, :vals, :ndel, :count, :age, :idxfloor, :maxprobe)
```

File: src/Tables.jl [hunk 2]
```diff
@@ -159,7 +159,7 @@ isrowtable(::Type{<:DictRows}) = true
 schema(x::DictRows) = nothing
 
 # Dict of AbstractVectors for Tables.columns
-const DictColumns = AbstractDict{K, V} where {K <: Union{Symbol, String}, V <: AbstractVector}
+const DictColumns = D where {D <: Union{AbstractDict{<:AbstractString, <:AbstractVector}, AbstractDict{Symbol, <:AbstractVector}}}
```
Comment (bkamins) [inline src/Tables.jl]: Note that this changes additionally previous behavior (which I think was incorrect - now `false` will be returned):
```
julia> Tables.istable(Dict{Union{String, Symbol}, Vector})
true
```

**Comment by bkamins** [comment]
I have left notes in places that change previous behavior.

File: test/runtests.jl [hunk 3]
```diff
@@ -900,3 +900,34 @@ end
...
+                                                                                                                                                
+    @test Set(Tables.columnnames(Dict(:a=>1, :b=>2))) == Set([:a, :b])
+    @test Set(Tables.columnnames(Dict("a"=>1, "b"=>2))) == Set([:a, :b])
+    @test Set(Tables.columnnames(Dict("a"=>1, SubString("b")=>2))) == Set([:a, :b])
+    @test Set(Tables.columnnames(Dict(SubString("a")=>1, SubString("b")=>2))) == Set([:a, :b])
+
+    @test Tables.getcolumn(Dict(:a=>1, :b=>2), 1) == 1
+    @test Tables.getcolumn(Dict("a"=>1, "b"=>2), 2) == 1
+    @test Tables.getcolumn(Dict("a"=>1, SubString("b")=>2), 2) == 1
+    @test Tables.getcolumn(Dict(SubString("a")=>1, SubString("b")=>2), 2) == 1
```
Comment (bkamins) [inline test/runtests.jl]:
```suggestion
    @test Tables.getcolumn(Dict("a"=>1, "b"=>2), 1 + (Int === Int64)) == 1
    @test Tables.getcolumn(Dict("a"=>1, SubString("b")=>2), 1 + (Int === Int64)) == 1
    @test Tables.getcolumn(Dict(SubString("a")=>1, SubString("b")=>2), 1 + (Int === Int64)) == 1
```

File: src/Tables.jl (hunk 2 above)
Comment (nalimilan) [reply src/Tables.jl]: Isn't this just:
```suggestion
const DictColumns = Union{<:AbstractDict{<:AbstractString, <:AbstractVector}, <:AbstractDict{Symbol, <:AbstractVector}}}
```

Comment (bkamins) [reply src/Tables.jl]: It is :)

**Review by nalimilan — APPROVED** [review]

File: src/Tables.jl [hunk 4]
```diff
@@ -159,7 +159,7 @@ isrowtable(::Type{<:DictRows}) = true
 schema(x::DictRows) = nothing
 
 # Dict of AbstractVectors for Tables.columns
-const DictColumns = AbstractDict{K, V} where {K <: Union{Symbol, String}, V <: AbstractVector}
+const DictColumns = Union{<:AbstractDict{<:AbstractString, <:AbstractVector}, <:AbstractDict{Symbol, <:AbstractVector}}}
```
Comment (nalimilan) [inline src/Tables.jl]:
```suggestion
const DictColumns = Union{<:AbstractDict{<:AbstractString, <:AbstractVector}, <:AbstractDict{Symbol, <:AbstractVector}}
```

**Review by quinnj — APPROVED** [review]

**Pull request merged**